### DISCUSSION TITLE
Computing resources seem very large

### DISCUSSION THREAD

**Discussion by niubiworker** [NanoComp/meep#2651] [Q&A]
I want to calculate the loss of a waveguide, but it seems that the resources for this calculation are very large. I have 16 processes and 128Gb of memory, but at a resolution of 30, there is still an error. I don't know if there is a problem with my program settings.
```
import` meep as mp
import numpy as np
import matplotlib.pyplot as plt
from meep.materials import LiNbO3

a = 0.775 
h = 0.3 
r = 0.315*a 
p_num = 50 
sx = 60
sy = 10*h
sz = 20
pad = 4
resolution = 40
courant = 0.1
geometry = [mp.Block(center=mp.Vector3(0,0,0),size=mp.Vector3(mp.inf,h,sz-2*pad),material=LiNbO3)]

wavel = 1.55
pml = 0.5*wavel
fcen = 1/wavel
df = 0.1
nfreq = 10 

sources = [mp.Source(mp.GaussianSource(fcen,fwidth=df),center=mp.Vector3(-0.5*sx + pml,0,0),
                     component=mp.Ey,size=mp.Vector3(0,h,np.sqrt(3)*a))]
sim = mp.Simulation(cell_size=mp.Vector3(sx,sy,sz),resolution=resolution,
                    boundary_layers=[mp.Absorber(pml)],sources=sources,
                    geometry=geometry,Courant=courant)
refl_fr = mp.FluxRegion(center=mp.Vector3(-0.5*sx+pml+0.5,0,0),direction=mp.X,
                        size=mp.Vector3(0,2*h,2*np.sqrt(3)*a))
refl = sim.add_flux(fcen,df,nfreq,refl_fr)
tran_fr = mp.FluxRegion(center=mp.Vector3(0.5*sx-pml,0,0),direction=mp.X,
                        size=mp.Vector3(0,2*h,a*np.sqrt(3)))
tran = sim.add_flux(fcen,df,nfreq,tran_fr)

pt = mp.Vector3(0.5*sx-pml-0.5,0,0)
sim.run(until_after_sources=mp.stop_when_fields_decayed(50,mp.Ey,pt,1e-3))
straight_refl_data = sim.get_flux_data(refl)
straight_tran_flux = mp.get_fluxes(tran)

sim.reset_meep()

y_num = 6
up_boundary = y_num*np.sqrt(3)*a
geometry = [mp.Block(center=mp.Vector3(0,0,0),size=mp.Vector3(sx,h,sz-2*pad),material=LiNbO3)]
for i in np.linspace(-2*p_num*a,2*p_num*a,p_num*4+1):
    for j in np.linspace(-up_boundary,up_boundary,2*y_num+1):
        if j != 0:
            geometry += [mp.Cylinder(center=mp.Vector3(i, 0, j, ), height=mp.inf, material=mp.air, radius=r,axis=mp.Vector3(y=1))]
            geometry += [mp.Cylinder(center=mp.Vector3(i + 0.5 * a, 0, np.sqrt(3) / 2 * a + j, ), height=mp.inf, material=mp.air,radius=r,axis=mp.Vector3(y=1))]
            geometry += [mp.Cylinder(center=mp.Vector3(i + 0.5 * a, 0, -np.sqrt(3) / 2 * a + j, ), height=mp.inf, material=mp.air,radius=r,axis=mp.Vector3(y=1))]
sim = mp.Simulation(resolution=resolution,
                    cell_size=mp.Vector3(sx,sy,sz),
                    geometry=geometry,
                    boundary_layers=[mp.Absorber(pml)],
                    sources=sources)

ref1 = sim.add_flux(fcen,df,nfreq,refl_fr)
tran_fr = mp.FluxRegion(center=mp.Vector3(-0.5*sx + pml+p_num*a,0,0),direction=mp.X,
                        size=mp.Vector3(0,2*h,a*np.sqrt(3)))
tran = sim.add_flux(fcen,df,nfreq,tran_fr)
sim.load_minus_flux_data(ref1,straight_refl_data)
sim.run(until_after_sources=mp.stop_when_fields_decayed(50, mp.Ez, pt, 1e-3))
pc_ref1_flux = mp.get_fluxes(ref1)
pc_tran_flux = mp.get_fluxes(tran)

flux_freqs = mp.get_flux_freqs(ref1)

wl = []
Rs = []
Ts = []
for i in range(nfreq):
    wl = np.append(wl, 1 / flux_freqs[i])
    Rs = np.append(Rs, -pc_ref1_flux[i] / straight_tran_flux[i])
    Ts = np.append(Ts, pc_tran_flux[i] / straight_tran_flux[i])

plt.plot(wl, Rs, "bo-", label="reflectance")
plt.plot(wl, Ts, "ro-", label="transmittance")
plt.plot(wl, 1 - Rs - Ts, "go-", label="loss")
plt.xlim(min(wl),max(wl))
plt.xlabel("wavelength (μm)")
plt.legend(loc="upper right")
plt.show()
```
The process of running the program and the errors encountered are in this file
[test2-0.1-30.txt](https://github.com/NanoComp/meep/files/12734790/test2-0.1-30.txt)
I would like to know how I should optimize next to obtain the calculation results of the 3D waveguide

**Comment by smartalecH** [comment]
I wouldn't use an FDTD solver to compute waveguide loss (especially for a straight waveguide). 

Just use an eigenmode solver. 

But in reality, your fab process is going to determine the actual loss. You'll get results very different from simulation due to things like sidewall roughness. So it's best to pull these numbers from metrology.

**Reply by niubiworker** [reply]
May I ask if there are any Python examples or links that I can learn from? Because I saw only the FDTD method for curved waveguides in the examples of reflection/transmission/loss, and did not see any examples of using the eigenmode solver. Thank you.

**Reply by smartalecH** [reply]
Meep isn't an eigenmode solver. You can use MPB, but that doesn't work with lossy materials.

You can try a variety of other python mode solvers. Tidy3D is a great one (that's also free and open source): https://www.flexcompute.com/tidy3d/examples/notebooks/ModeSolver/

**Reply by simbilod** [reply]
Like Alec says, an eigenmode solver is your best bet. 

If you are truly certain loss is coming from sidewall scattering and/or absorption from dangling bonds, and you have some metrology data to fit, then maybe a FEM eigensolver can allow you to incorporate a thin absorption region more easily. See for instance https://github.com/HelgeGehring/femwell/pull/68

But sidewall loss might not be explaining all the loss you see (could also come from leakage, metal/doping proximity, etc) so this can also be misleading.

**Comment by stevengj** [comment]
> You can use MPB, but that doesn't work with lossy materials.

In almost any practical dielectric waveguide, the absorption loss should be small enough to compute with perturbation theory.  That is compute the dispersion relation for the lossless problem (using MPB or other tools), then compute the absorption loss with the perturbation theory formula (compute the fraction of ε|E|^2 in the lossy material, multiply by Im n / Re n, then divide by the group velocity … the result is Im β, the imaginary part of the propagation constant, to first order in Im n / Re n)
